derive_client/utils/w3.py

```diff
@@ -163,3 +164,61 @@ def estimate_fees(w3, percentiles: list[int], blocks=20, default_tip=10_000):
         fee_estimations.append({"maxFeePerGas": max_fee, "maxPriorityFeePerGas": priority_fee})
 
     return fee_estimations
+
+
+def log_matches_topics(log: AttributeDict, topics: list[str | None]) -> bool:
+    for i, topic in enumerate(topics):
+        if topic is None:
+            continue
+        if i >= len(log["topics"]) or log["topics"][i].to_0x_hex() != topic:
+            return False
+    return True
+
+
+def iter_events(
+    w3: Web3,
+    event_filter: EventFilter,
+    max_block_range: int = 10_000,
+    poll_interval: float = 5.0,
+    timeout: float | None = None,
+) -> Generator[AttributeDict, None, None]:
+    """Stream matching logs over a fixed or live block window. Optionally raises TimeoutError."""
+
+    if (cursor := event_filter.from_block) == "latest":
+        cursor = w3.eth.block_number
+
+    start_block = cursor
+    fixed_ceiling = None if event_filter.to_block == "latest" else event_filter.to_block
+
+    filter_params = event_filter.model_dump(by_alias=True)
+    deadline = None if timeout is None else time.time() + timeout
+    while True:
+        if deadline and time.time() > deadline:
+            msg = f"Timed out waiting for events after scanning blocks {start_block}-{cursor}"
+            raise TimeoutError(f"{msg}: {event_filter}")
+        upper = fixed_ceiling or w3.eth.block_number
+        if cursor <= upper:
+            end = min(upper, cursor + max_block_range - 1)
+            filter_params["fromBlock"] = hex(cursor)
+            filter_params["toBlock"] = hex(end)
```

Comment:
I didn't see this error before all day 
```
TxResult(tx_hash='', tx_receipt=None, exception=ValueError({'message': 'invalid block range params', 'code': -32000}))```

until I did just now! Not sure this hex conversion of its fixes it, but hopefully. I mean, how am I supposed to know what version these RPC providers are using / running?

TODO: add a retry mechanism on failures (and logging of exceptions for debugging)